src/book/02-system/01-foundations/07-director.mdx

```diff
@@ -0,0 +1,231 @@
+This means the walk engine could have a `Leaving` reaction that caused this state so other modules know it can use this to be standing before they execute their actions.
+
+```cpp
+on<Leaving<Walk>, Causing<Stability, Stability::STANDING>().then(...);
+```
+
+Other providers will take advantage of this in two ways.
+If the other provider has higher priority than this one, it will force this provider group to switch to a `Provides` or `Leaving` that has a causing that satisfies the when condition it is limited on.
+If the other provider has lower priority than this one, it will only be able to force this provider to change to another `Provides` statement and not a leavign one.
+
+## Tasks
+
+Tasks are the jobs that should be executed using the providers.
+Tasks are implemented using NUClears `emit` along with a custom scope `emit<Task>`.
+
+### Root level tasks
+
+When a task is emitted from a reaction which is not a provider, that task is considered a root level task.
+These root level tasks are all considered siblings within the priority comparision system at the root level of the graph.
+When a root level task is emitted with 0 priority it is removed from the list of tasks.
+
+### Subtasks
+
+A subtask is a task that is emitted by a provider.
+When multiple tasks are emitted by a provider, each of these are sibling tasks in the director graph.
+
+### Priority
+
+Priority in the Director is considered based on the closest common anscestor.
+For root tasks this will be the root element.
+Once the closest common anscestor is determined, the priority of the branch we are on is compared to the priority of the branch of the other task.
+The winner gets to use the resources and potentially evict the other task to wait in the queue for the provider group.
+
+Tasks can also be emitted as optional tasks, if when finding a common anscestor between two tasks, an optional task is found in the parantage of a task before the common anscestor that task is considered "optional".
+If one task has an optional parentage and the other does not then the optional task will automatically lose.
+If both have optional tasks in their perantage, then the tasks will be compared using their common anscestor as normal as if the optional was not there.
+This means the higher priority root tasks optional elements are more important than a lower priority tasks optional elements.
+
+### Done tasks
+
+A Done task is a special task that can be emitted by a provider to signal to its parent that it has completed the task that was assigned to it.
+The provider group that created this task will then be re-executed with the knowledge that it was triggered as a done event from below.
+If a root task emits a `Done` task then it will be removed from the Director's root tasks as if it were emitted with a priority of 0.
+
+## The Director Graph
+
+TODO algorithm development in progress
+
+## Behaviours Layout
```

Comment:
About this whole section - so all of these do become providers? Webots, HardwareIO, QuinticWalk, etc? This is a complete codebase wide change? This almost feels like NUClear 2.0 but maybe I think it does more than what it really does because I don't understand it properly.

Reply:
It is a fairly codebase wide change, but a lot of it will just be replacing things that were previously on<Trigger<X>> with on<Provides<X>> so that the director can choose when to run them.

Some of the modules will get smaller because of this and some will likely need some retooling and splitting up.

At some point when we want to start working on the modules I intend to make a list of all the modules I see currently in the codebase so people can start building and testing them.

Best part about the director is you can start with any task as a root level task which will make it super easy to test things.
No more of this weird subsumption registration so you can get PS3 walk going. PS3 walk would just emit a walk command as a root level task and update it.

Reply:
Sounds good. I'm wondering how we can make the change incrementally without having a huge PR that causes major merge conflicts for every existing branch. Will this also change how we write roles?

Reply:
I'd recommend doing it in modules from the bottom up.
So write the servo module
Then the IK module
Then the walk module
etc

Reply:
Will this also change how we write roles?